db/migrate/20200617014509_add_subscriber_email_index_to_user_subscriptions.rb

```diff
@@ -0,0 +1,45 @@
+class AddSubscriberEmailIndexToUserSubscriptions < ActiveRecord::Migration[6.0]
+  disable_ddl_transaction!
+
+  def up
+    if index_exists?(:user_subscriptions, %i[subscriber_id user_subscription_sourceable_id user_subscription_sourceable_type])
+      remove_index :user_subscriptions,
+                   column: %i[subscriber_id user_subscription_sourceable_id user_subscription_sourceable_type],
+                   algorithm: :concurrently
+    end
+
+    unless index_exists?(:user_subscriptions, :subscriber_email)
+      add_index :user_subscriptions,
+                :subscriber_email,
+                algorithm: :concurrently
+    end
+
+    unless index_exists?(:user_subscriptions, %i[subscriber_id subscriber_email user_subscription_sourceable_type user_subscription_sourceable_id])
+      add_index :user_subscriptions,
+                %i[subscriber_id subscriber_email user_subscription_sourceable_type user_subscription_sourceable_id],
+                name: "index_subscriber_id_and_email_with_user_subscription_source",
+                unique: true,
+                algorithm: :concurrently
+    end
+  end
+
+  def down
+    unless index_exists?(:user_subscriptions, %i[subscriber_id user_subscription_sourceable_id user_subscription_sourceable_type])
+      add_index :user_subscriptions,
+                %i[subscriber_id user_subscription_sourceable_id user_subscription_sourceable_type],
+                unique: true,
+                name: :index_on_subscriber_id_user_subscription_sourceable_type_and_id,
+                algorithm: :concurrently
+    end
+
+    if index_exists?(:user_subscriptions, :subscriber_email)
+      remove_index :user_subscriptions, column: :subscriber_email, algorithm: :concurrently
+    end
+
+    if index_exists?(:user_subscriptions, %i[subscriber_id subscriber_email user_subscription_sourceable_type user_subscription_sourceable_id])
+      remove_index :user_subscriptions,
+                   column: %i[subscriber_id subscriber_email user_subscription_sourceable_type user_subscription_sourceable_id],
+                   algorithm: :concurrently
+    end
+  end
+end
```

Comment:
I might be missing something completely obvious here, but why do we need to explicitly check whether or not each index exists before removing it (or even adding it)? I would imagine that we would know when they would exist based on whether the `up` or `down` method are being run; is this for idempotency's sake?

Reply:
Just for idempotency 💯 

We've had a couple of issues in production in the past. A new PR with a migration goes out, the migration run, but the deploy fails for some reason. The PR is rolled back, a fix is pushed, but then it tries to run the same migration again and the index exists already, throwing an error, causing another problem.

We should probably document this somewhere now that I think about it...but the "new rule of thumb" has become to make the migrations idempotent to avoid that headache.

Reply:
Ah okay, makes total sense :)

I basically asked the question and then answered it for myself, haha